Documentation for example setups of virtual address / load balancing … (#2102)
kubernetes/kubeadm: docs/ha-considerations.md

# High Availability Considerations

> 

Reviewer: I think we have to add a disclaimer on top of this documents stating that:
- this page contains a collection of community provided High Availability Considerations
- if a user find something incomplete, not clear, or if the user is willing to document alternative solutions, contributions are well accepted.
- being this a community provided list of consderations, the best place to ask support or information about this page is the kubeadm channel where most of the contributors are usually active

@neolit123 something else to add to clarify "the contract/the support matrix" around this page?

Author: Thanks, I have created a disclaimer with what you proposed.

Reviewer 2: > @neolit123 something else to add to clarify "the contract/the support matrix" around this page?

not sure what you mean?

Reviewer: @neolit123 I was asking if you want to add/change something in the disclaimer

Reviewer 2: ok, the disclaimer seems fine for now.

## Overview

When setting up a production cluster, high availability (the cluster's ability to remain operational even if some control plane or worker nodes fail) is usually a requirement. For worker nodes, assuming that there are enough of them, this is part of the very cluster functionality. However redundancy of control plane nodes and `etcd` instances needs to be catered for when planning and setting up a cluster.

`kubeadm` supports setting up of multi control plane and multi `etcd` clusters (see [Creating Highly Available clusters with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/) for a step-by-step guide). Still there are some aspects to consider and set up which are not part of Kubernetes itself and hence not covered in the project documentation. This document provides some additional information and examples useful when planning and bootstrapping HA clusters with `kubeadm`.

> <!-- TOC -->

Reviewer: I think we can move this before overview

Author: Maybe not if there's already the disclaimer on top? Would IMO not quite fit anymore.

Reviewer: I just find weird to see the topic list not at the beginning of the doc, but it is only a nit

Author: That seems a matter of taste to me. I don't have any strong feelings here, so if you guys would like it changed I'm fine with it. In that case please give me an indication how you would like it. 

As of now the structure is: main header, disclaimer, then section Overview with a brief introduction text and the TOC, then the rest. I find this OK, but there's surely various other ways of structuring it.

    - [keepalived and haproxy](#keepalived-and-haproxy)
    - [kube-vip](#kube-vip)
    - [Bootstrap the cluster](#bootstrap-the-cluster)

## Options for Software Load Balancing

When setting up a cluster with more than one control plane, higher availability can be achieved by putting the API Server instances behind a load balancer and using the `--control-plane-endpoint` option when running `kubeadm init` for the new cluster to use it.

Of course, the load balancer itself should be highly available, too. This is usually achieved by adding redundancy to the load balancer. In order to do so, a cluster of hosts managing a virtual IP is set up with each host running an instance of the load balancer, so that always the load balancer on the host currently holding the vIP will be used while the others are on standby.

In some environments, like in data centers with dedicated load balancing components (provided e.g. by some cloud-providers), this functionality may already be available. If it is not, user-managed load balancing can be used. In that case some preparation is necessary before bootstrapping a cluster.

> Since this is infrastructure Kubernetes runs on and hence is not part of Kubernetes or `kubeadm`, this must be taken care of separately. In the following sections we give examples that have been working for some people while of course there are potentially dozens of other possible configurations.

Reviewer: ```suggestion
Since this is not part of Kubernetes nor in the scope of `kubeadm`, this must be taken care of separately. In the following sections, we give examples that have been working for some people while of course there are potentially dozens of other possible configurations.
```

Author: (y)

# High Availability Considerations

## Overview

> When setting up a productive cluster, high availability, i.e. the cluster's ability to remain operational even if some control plane or worker nodes fail, is usually a requirement. For worker nodes, assuming that there are enough of them, this is part of the very cluster functionality; however redundancy of control plane nodes and `etcd` instances needs to be catered for when planning and setting up a cluster.

Reviewer 2: productive -> production
as users are more used to the later.

---

> high availability, i.e. the cluster's ability to remain operational even if some control plane or worker nodes fail,

i'd change this to:

> high availability (the cluster's ability to remain operational even if some control plane or worker nodes fail), ...

k/website recommends to not use latin abbreviations in docs, although we don't follow this strictly on our side.

---

`functionality; however redundancy` -> `functionality. However redundancy`

> `kubeadm` supports setting up of multi control plane and multi `etcd` clusters (see [Creating Highly Available clusters with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/), part of the official Kubernetes documentation, for a step-by-step guide). However there are some more things to consider and set up which are not part of Kubernetes itself and hence not covered there. This document provides some additional information and examples useful when planning and bootstrapping HA clusters with `kubeadm`.

Reviewer 2: i'd remove
`, part of the official Kubernetes documentation,`

---

`However there are some more things` -> `However there are some aspects...`

---

`not covered there.` -> `not covered in the project documentation.`

---

<!-- TOC -->

- [High Availability Considerations](#high-availability-considerations)
    - [Overview](#overview)
    - [Options for Software Load Balancing](#options-for-software-load-balancing)
        - [keepalived and haproxy example](#keepalived-and-haproxy-example)
        - [kube-vip example](#kube-vip-example)

>         - [Bootstrap the cluster](#bootstrap-the-cluster)

Reviewer 2: as mentioned during the meeting this should be one level up.

## Options for Software Load Balancing

When setting up a cluster with more than one control plane, one will typically put the API Server instances behind a load balancer and use the `--control-plane-endpoint` option when running `kubeadm init` for the new cluster to use it.

> Of course, the load balancer itself should be highly available, too. This is usually achieved by setting up more load balancers in a cluster of hosts managing a virtual IP, so that always the load balancer on the host currently holding the vIP will be used.

Reviewer 2: > This is usually

do we know for sure that this is the more common way out there?

> setting up more load balancers in a cluster of hosts managing a virtual IP

i don't think that if we deploy multiple keepalived instances on the nodes we should say that there are multiple loadbalancer for that cluster. it's still a single load balancer with multiple instances.

Author: > > This is usually
> 
> do we know for sure that this is the more common way out there?

I'm rewording this a bit to "higher availability can be achieved", that's more neutral. 

> > setting up more load balancers in a cluster of hosts managing a virtual IP
> 
> i don't think that if we deploy multiple keepalived instances on the nodes we should say that there are multiple loadbalancer for that cluster. it's still a single load balancer with multiple instances.

Rewording this, too. Let's see if it's better.

> In some environments, like in data centers with dedicated load balancing components, this functionality may already be available. If it is not, software load balancing can be used. In that case some preparation is necessary before bootstrapping a cluster.

Reviewer 2: should we mention cloud providers here too?
data centers are less common from the kubeadm survey.

> software load balancing can be used.

do you mean that in data centers the LBs can be hardware?
perhaps we should instead, differentiate between PaaS / IaaS manged LBs and user managed LBs here.

Author: > perhaps we should instead, differentiate between PaaS / IaaS manged LBs and user managed LBs here.

Good point.

Since this is infrastructure Kubernetes runs on and hence is not part of Kubernetes or `kubeadm`, this must be taken care of separately. In the following sections we give examples that have been working for some people while of course there are potentially dozens of other possible configurations.

> ## keepalived and haproxy example

Reviewer 2: i'd remove `example` from the title here.

For providing load balancing from a virtual IP the combination [keepalived](https://www.keepalived.org) and [haproxy](https://www.haproxy.com) has been around for a long time and can be considered well-known and well-tested:
- The `keepalived` service provides a virtual IP managed by a configurable health check. Due to the way the virtual IP is implemented, all the hosts between which the virtual IP is negotiated need to be in the same IP subnet.
- The `haproxy` service can be configured for simple stream-based load balancing thus allowing TLS termination to be handled by the API Server instances behind it.

This combination can be run either as services on the operating system or as static pods on the control plane hosts. The service configuration is identical for both cases.

### keepalived configuration

The `keepalived` configuration consists of two files: the service configuration file and a health check script which will be called periodically to verify that the node holding the virtual IP is still operational.

> The files are assumed to reside in a _/etc/keepalived_ directory. Note that however some Linux distributions may keep them elsewhere. The following configuration has been successfully used with `keepalived` version 1.3.5:

Reviewer 2: _/etc/keepalived_ ->  `/etc/keepalived` (with code back quotes)

Note that however -> Note that

```bash
! /etc/keepalived/keepalived.conf
! Configuration File for keepalived
global_defs {
    router_id LVS_DEVEL
}
vrrp_script check_apiserver {
  script "/etc/keepalived/check_apiserver.sh"
  interval 3
  weight -2
  fall 10
  rise 2
}

vrrp_instance VI_1 {
    state ${STATE}
    interface ${INTERFACE}
    virtual_router_id ${ROUTER_ID}
    priority ${PRIORITY}
    authentication {
        auth_type PASS
        auth_pass ${AUTH_PASS}
    }
    virtual_ipaddress {
        ${APISERVER_VIP}
    }
    track_script {
        check_apiserver
    }
}
```

> There are some placeholders in `bash` Variable style to fill in:

Reviewer 2: Variable -> variable

> - `${STATE}` is `MASTER` for one and `BACKUP` for all other hosts (hence the virtual IP will initially be assigned to the `MASTER`).

Reviewer 2: "(hence the virtual IP will initially be assigned to the `MASTER`)" ->  ", hence the virtual IP will initially be assigned to the `MASTER`"

- `${INTERFACE}` is the network interface taking part in the negotiation of the virtual IP, e.g. `eth0`.
- `${ROUTER_ID}` should be the same for all `keepalived` cluster hosts while unique amongst all clusters in the same subnet. Many distros pre-configure its value to `51`.
- `${PRIORITY}` should be higher on the master than on the backups. Hence `101` and `100` respectively will suffice.
- `${AUTH_PASS}` should be the same for all `keepalived` cluster hosts, e.g. `42`
- `${APISERVER_VIP}` is the virtual IP address negotiated between the `keepalived` cluster hosts.

### haproxy configuration

> The `haproxy` configuration consists of one file: the service configuration file which is assumed to reside in a _/etc/haproxy_ directory. Note that however some Linux distributions may keep them elsewhere. The following configuration has been successfully used with `haproxy` version 2.1.3:

Reviewer 2: _/etc/haproxy_ -> `/etc/haproxy` (with code back quotes)

# apiserver frontend which proxys to the masters
#---------------------------------------------------------------------
frontend apiserver
    bind *:${APISERVER_DEST_PORT}
    mode tcp
    option tcplog
    default_backend apiserver

#---------------------------------------------------------------------
# round robin balancing for apiserver
#---------------------------------------------------------------------
backend apiserver
    option httpchk GET /healthz
    http-check expect status 200
    mode tcp
    option ssl-hello-chk
    balance     roundrobin
        server ${HOST1_ID} ${HOST1_ADDRESS}:${APISERVER_SRC_PORT} check
        # [...]
```
Again, there are some placeholders in `bash` variable style to expand:
- `${APISERVER_DEST_PORT}` the port through which Kubernetes will talk to the API Server.
- `${APISERVER_SRC_PORT}` the port used by the API Server instances
- `${HOST1_ID}` a symbolic name for the first load-balanced API Server host
- `${HOST1_ADDRESS}` a resolvable address (DNS name, IP address) for the first load-balanced API Server host
- additional `server` lines, one for each load-balanced API Server host

> ### Option 1: Run the services on the operating system

Reviewer 2: should this and Option 2 be `####`?

Author: > should this and Option 2 be `####`?

I have decided to put the two examples on level ## to avoid too deep nesting. Hence here level ### would be corrrect. 
I think this makes sense because it avoids too deep nesting.

In order to run the two services on the operating system, the respective distribution's package manager can be used to install the software. This can make sense if they will be running on dedicated hosts not part of the Kubernetes cluster.

On a RedHat-style Linux distribution this will most likely look like this:
```

> # yum install keepalived haproxy

Reviewer 2: i will remove the distro specific examples for `yum`

```
Having now installed the abovementioned configuration, the services can be enabled and started. On a recent RedHat-based system, `systemd` will be used for this:
```
# systemctl enable haproxy --now

> # systemctl enable keepalived --now

Reviewer 2: given kubeadm already requires systemd on Linux, i think it's fine to leave the systemd examples

```
With the services up, now the Kubernetes cluster can be bootstrapped using `kubeadm init` (see [below](#bootstrap-the-cluster)).

> ### Option 2: Run the services as static pods

Reviewer 2: should this be `####`

> If `keepalived` and `haproxy` will be running on the control plane nodes they can be configured to run as static pods. All that is necessary here is placing respective manifest files in the _/etc/kubernetes/manifests_ directory before bootstrapping the cluster. During the bootstrap process, `kubelet` will bring the processes up, so that the cluster can use them while starting. This is an elegant solution, in particular with the setup described under [Stacked control plane and etcd nodes](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/#stacked-control-plane-and-etcd-nodes), where the `etcd` cluster, too, will be running as static pods on the control plane nodes.

Reviewer 2: please use ` ...` for folders instead of "_ ... _"
[1]

Reviewer 2: i'd remove the last part ` , where the `etcd` cluster, too, will be running as static pods on the control plane nodes.`.
with the work on etcdadm we'd like to delegate these steps to it.

> For this setup, two manifest files need to be created in _/etc/kubernetes/manifests_ (create the directory first).

Reviewer 2: please use ` ...` for folders instead of "_ ... _"
[1]

> The manifest for `keepalived`, _/etc/kubernetes/manifests/keepalived.yaml_:

Reviewer 2: [1]

```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  name: keepalived
  namespace: kube-system
spec:
  containers:

>   - image: osixia/keepalived:1.3.5-1

Reviewer 2: i haven't seen the name osixia before.
do we have a more "official" image maintained at docker hub?

Author: > i haven't seen the name osixia before.
> do we have a more "official" image maintained at docker hub?

I did not find anything. While there is a Dockerfile distributed there isn't anything labeled as "official" or even "certified" on Docker Hub.

Reviewer 2: ok, understood. maybe someone else will have ideas.

    name: keepalived
    resources: {}
    securityContext:
      capabilities:
        add:
        - NET_ADMIN
        - NET_BROADCAST
        - NET_RAW
    volumeMounts:
    - mountPath: /usr/local/etc/keepalived/keepalived.conf
      name: config
    - mountPath: /etc/keepalived/check_apiserver.sh
      name: check
  hostNetwork: true
  volumes:
  - hostPath:
      path: /etc/keepalived/keepalived.conf
    name: config
  - hostPath:
      path: /etc/keepalived/check_apiserver.sh
    name: check
status: {}
```

> The manifest for `haproxy`, _/etc/kubernetes/manifests/haproxy.yaml_:

Reviewer 2: [1]

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: haproxy
  namespace: kube-system
spec:
  containers:

>   - image: haproxy:2.1.3-alpine

Reviewer 2: wondering if alpine should not be recommended due to it's musl usage and related DNS issues.

Author: > wondering if alpine should not be recommended due to it's musl usage and related DNS issues.

Good point.

    name: haproxy
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: localhost
        path: /healthz
        port: ${APISERVER_DEST_PORT}
        scheme: HTTPS
    volumeMounts:
    - mountPath: /usr/local/etc/haproxy/haproxy.cfg
      name: haproxyconf
      readOnly: true
  hostNetwork: true
  volumes:
  - hostPath:
      path: /etc/haproxy/haproxy.cfg
      type: FileOrCreate
    name: haproxyconf
status: {}
```

> Note that here again a placeholder needs to be filled in: `${APISERVER_DEST_PORT}` needs to hold the same value as in _/etc/haproxy/haproxy.cfg_ (see above).

Reviewer 2: [1]

This combination has been successfully used with the versions used in the example. Other versions might work as well or may require changes to the configuration files.

With the services up, now the Kubernetes cluster can be bootstrapped using `kubeadm init` (see [below](#bootstrap-the-cluster)).

> ## kube-vip example

Reviewer 2: i'd remove `example` from the title.

As an alternative to the more "traditional" approach of `keepalived` and `haproxy`, [kube-vip](https://plndr.io/kube-vip/) implements both management of a virtual IP and load balancing in one service. Similar to option 2 above, `kube-vip` will be run as a static pod on the control plane nodes.

Like with `keepalived`, the hosts negotiating a virtual IP need to be in the same IP subnet. Similarly, like with `haproxy`, stream-based load-balancing allows TLS termination to be handled by the API Server instances behind it.

> The configuration file _/etc/kube-vip/config.yaml_ looks like this:

Reviewer 2: [1]

singleNode: false
startAsLeader: ${IS_LEADER}
interface: ${INTERFACE}
loadBalancers:
- name: API Server Load Balancer
  type: tcp
  port: ${APISERVER_DEST_PORT}
  bindToVip: false
  backends:
  - port: ${APISERVER_SRC_PORT}
    address: ${HOST1_ADDRESS}
  # [...]
```

The `bash` style placeholders to expand are these:
- `${ID}` the current host's symbolic name
- `${IPADDR}` the current host's IP address
- `${PEER1_ID}` a symbolic name for the first vIP peer
- `${PEER1_IPADDR}` IP address for the first vIP peer
- entries (`id`, `address`, `port`) for additional vIP peers can follow
- `${APISERVER_VIP}` is the virtual IP address negotiated between the `kube-vip` cluster hosts.
- `${IS_LEADER}` is `true` for exactly one leader and `false` for the rest
- `${INTERFACE}` is the network interface taking part in the negotiation of the virtual IP, e.g. `eth0`.
- `${APISERVER_DEST_PORT}` the port through which Kubernetes will talk to the API Server.
- `${APISERVER_SRC_PORT}` the port used by the API Server instances
- `${HOST1_ADDRESS}` the first load-balanced API Server host's IP address
- entries (`port`, `address`) for additional load-balanced API Server hosts can follow

> To have the service started with the cluster, now the manifest `kube-vip.yaml` needs to be placed in _/etc/kubernetes/manifests_ (create the directory first). It can be generated using the `kube-vip` docker image:

Reviewer 2: [1]

```
# docker run -it --rm plndr/kube-vip:0.1.1 /kube-vip sample manifest \
    | sed "s|plndr/kube-vip:'|plndr/kube-vip:0.1.1'|" \
    | sudo tee /etc/kubernetes/manifests/kube-vip.yaml
```

> The result, _/etc/kubernetes/manifests/kube-vip.yaml_, will look like this:

Reviewer 2: please add a new line after the previous code block.

also [1]

    name: kube-vip
    resources: {}
    securityContext:
      capabilities:
        add:
        - NET_ADMIN
        - SYS_TIME
    volumeMounts:
    - mountPath: /vip.yaml
      name: config
  hostNetwork: true
  volumes:
  - hostPath:
      path: /etc/kube-vip/config.yaml
    name: config
status: {}
```

With the services up, now the Kubernetes cluster can be bootstrapped using `kubeadm init` (see [below](#bootstrap-the-cluster)).

## Bootstrap the cluster

Now the actual cluster bootstrap as described in [Creating Highly Available clusters with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/) can take place.

Note that, if `${APISERVER_DEST_PORT}` has been configured to a value different from `6443` in the configuration above, `kubeadm init` needs to be told to use that port for the API Server. Assuming that in a new cluster port 8443 is used for the load-balanced API Server and a virtual IP with the DNS name `vip.mycluster.local`, an argument `--control-plane-endpoint` needs to be passed to `kubeadm` as follows:

```
# kubeadm init --control-plane-endpoint vip.mycluster.local:8443 [additional arguments ...]
```

> 

Reviewer 2: we can leave only one empty line after the last code block.
